File: crates/wasm/tests-node/vectors.mjs
// Cross-target conformance: the committed corpus, run through the real
// wasm-bindgen glue.
//
// This is deliberately Node rather than a pure-Rust wasm test. A `cargo test`
// under wasm32 exercises the Rust, but not the generated JavaScript that every
// browser actually calls — and that glue is where the interesting bugs are.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import init, { openEnvelope, openEnvelopeWithPassphrase, formatVersion } from "../pkg/sirna_wasm.js";

const here = path.dirname(fileURLToPath(import.meta.url));
const vectorDir = path.join(here, "../../../spec/vectors");

await init({ module_or_path: fs.readFileSync(path.join(here, "../pkg/sirna_wasm_bg.wasm")) });

const manifest = JSON.parse(fs.readFileSync(path.join(vectorDir, "vectors.json"), "utf8"));

if (manifest.format_version !== formatVersion()) {
  console.error(
    `corpus targets format ${manifest.format_version}, this build produces ${formatVersion()}`,
  );
  process.exit(1);
}

// The same fixed clock the generator used. A real clock would make expiry
// vectors start failing on their own.
const NOW = 1800000000n;

let passed = 0;
let failed = 0;

for (const v of manifest.vectors) {
  const envelope = fs.readFileSync(path.join(vectorDir, v.envelope_file));

  let outcome;
  try {
    const opened = v.passphrase
      ? openEnvelopeWithPassphrase(envelope, v.passphrase, NOW)
      : openEnvelope(envelope, v.mnemonic, NOW);
    outcome = { ok: true, opened };
  } catch (e) {
    outcome = { ok: false, code: String(e?.code ?? "?") };
  }

  if (v.expect === "ok") {
    if (!outcome.ok) {
      console.error(`FAIL ${v.id}: expected success, got code ${outcome.code}`);
      failed++;
      continue;
    }
    if (BigInt(outcome.opened.plaintext.length) !== BigInt(v.plaintext_len)) {
      console.error(
        `FAIL ${v.id}: length ${outcome.opened.plaintext.length} != ${v.plaintext_len}`,
      );
      failed++;
      continue;
    }
    passed++;
  } else {
    // Assert the numeric code, never the message. Codes are the contract that
    // holds three clients together; messages are UI copy.
    if (outcome.ok) {
      console.error(`FAIL ${v.id}: expected error ${v.expect}, but it opened`);
      failed++;
    } else if (outcome.code !== v.expect) {
      console.error(`FAIL ${v.id}: expected error ${v.expect}, got ${outcome.code}`);
      failed++;
    } else {
      passed++;
    }
  }
}

console.log(`wasm vectors: ${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);

File: android/tests-jvm/Vectors.kt
// Cross-target conformance for the Kotlin bindings, on a desktop JVM.
//
// No emulator, no Android SDK, no NDK, no phone. That is the whole point: the
// Android client's agreement with the format is provable long before the app
// exists, and long before any of the ten gigabytes of Android tooling is
// installed anywhere.
//
// A format that three clients each implement will drift. The drift is invisible
// until the day a file written on a phone will not open on a laptop, and by
// then it is in a release. This is the cheapest possible place to catch it.
//
// Mirrors crates/wasm/tests-node/vectors.mjs exactly, including the rule that
// negative cases assert the numeric error code and never the message — codes
// are the contract, messages are UI copy.

import uniffi.sirna_ffi.*
import java.io.File

// The same fixed clock the generator used. A real clock would make the expiry
// vectors start failing on their own.
const val NOW: ULong = 1800000000uL

/** Minimal field reader — enough for a flat manifest, and it keeps this test
 *  free of a JSON dependency that would have to be fetched and pinned. */
fun field(obj: String, name: String): String? {
    val key = "\"$name\""
    val at = obj.indexOf(key)
    if (at < 0) return null
    var i = obj.indexOf(':', at + key.length) + 1
    while (i < obj.length && obj[i].isWhitespace()) i++
    if (obj[i] != '"') {
        val end = obj.indexOfFirst(i) { it == ',' || it == '}' }
        return obj.substring(i, end).trim()
    }
    val sb = StringBuilder()
    i++
    while (obj[i] != '"') {
        if (obj[i] == '\\') i++
        sb.append(obj[i]); i++
    }
    return sb.toString()
}

fun String.indexOfFirst(from: Int, pred: (Char) -> Boolean): Int {
    var i = from
    while (i < length && !pred(this[i])) i++
    return i
}

/** Split the manifest's `vectors` array into one string per object. */
fun splitObjects(json: String): List<String> {
    val start = json.indexOf("\"vectors\"")
    val out = mutableListOf<String>()
    var depth = 0
    var begin = -1
    var i = json.indexOf('[', start)
    while (i < json.length) {
        when (json[i]) {
            '{' -> { if (depth == 0) begin = i; depth++ }
            '}' -> { depth--; if (depth == 0) out.add(json.substring(begin, i + 1)) }
            ']' -> if (depth == 0) return out
        }
        i++
    }
    return out
}

fun main() {
    val root = File(System.getProperty("sirna.vectors") ?: "spec/vectors")
    val manifest = File(root, "vectors.json").readText()

    val declared = field(manifest.substring(0, manifest.indexOf("\"vectors\"")), "format_version")!!
    if (declared.toUByte() != formatVersion()) {
        System.err.println("corpus targets format $declared, this build produces ${formatVersion()}")
        kotlin.system.exitProcess(1)
    }

    var passed = 0
    var failed = 0

    for (v in splitObjects(manifest)) {
        val id = field(v, "id")!!
        val expect = field(v, "expect")!!
        val passphrase = field(v, "passphrase")
        val mnemonic = field(v, "mnemonic")!!
        val plaintextLen = field(v, "plaintext_len")!!.toLong()
        val envelope = File(root, field(v, "envelope_file")!!).readBytes()

        var opened: Opened? = null
        var code: String? = null
        try {
            opened = if (passphrase != null) {
                openEnvelopeWithPassphrase(envelope, passphrase, NOW)
            } else {
                openEnvelope(envelope, mnemonic, NOW)
            }
        } catch (e: SirnaException.Failed) {
            code = e.code.toString()
        }

        if (expect == "ok") {
            when {
                opened == null ->
                    { System.err.println("FAIL $id: expected success, got code $code"); failed++ }
                opened.plaintext.size.toLong() != plaintextLen ->
                    { System.err.println("FAIL $id: length ${opened.plaintext.size} != $plaintextLen"); failed++ }
                else -> passed++
            }
        } else {
            when {
                opened != null ->
                    { System.err.println("FAIL $id: expected error $expect, but it opened"); failed++ }
                code != expect ->
                    { System.err.println("FAIL $id: expected error $expect, got $code"); failed++ }
                else -> passed++
            }
        }
    }

    println("kotlin vectors: $passed passed, $failed failed")
    kotlin.system.exitProcess(if (failed == 0) 0 else 1)
}
